src/cargo/core/source/source_id.rs

```diff
@@ -11,70 +10,26 @@ use serde::ser;
 use serde::de;
 use url::Url;
 
-use core::{Package, PackageId, Registry};
 use ops;
 use sources::git;
 use sources::{PathSource, GitSource, RegistrySource, CRATES_IO};
 use sources::DirectorySource;
-use util::{Config, CargoResult, ToUrl};
+use util::{Config, ConfigValue as CV, CargoResult, ToUrl};
 
-/// A Source finds and downloads remote packages based on names and
-/// versions.
-pub trait Source: Registry {
-    fn source_id(&self) -> &SourceId;
-
-    /// The update method performs any network operations required to
-    /// get the entire list of all names, versions and dependencies of
-    /// packages managed by the Source.
-    fn update(&mut self) -> CargoResult<()>;
-
-    /// The download method fetches the full package for each name and
-    /// version specified.
-    fn download(&mut self, package: &PackageId) -> CargoResult<Package>;
-
-    /// Generates a unique string which represents the fingerprint of the
-    /// current state of the source.
-    ///
-    /// This fingerprint is used to determine the "fresheness" of the source
-    /// later on. It must be guaranteed that the fingerprint of a source is
-    /// constant if and only if the output product will remain constant.
-    ///
-    /// The `pkg` argument is the package which this fingerprint should only be
-    /// interested in for when this source may contain multiple packages.
-    fn fingerprint(&self, pkg: &Package) -> CargoResult<String>;
-
-    /// If this source supports it, verifies the source of the package
-    /// specified.
-    ///
-    /// Note that the source may also have performed other checksum-based
-    /// verification during the `download` step, but this is intended to be run
-    /// just before a crate is compiled so it may perform more expensive checks
-    /// which may not be cacheable.
-    fn verify(&self, _pkg: &PackageId) -> CargoResult<()> {
-        Ok(())
-    }
+/// Unique identifier for a source of packages.
+#[derive(Clone, Eq, Debug)]
+pub struct SourceId {
+    inner: Arc<SourceIdInner>,
 }
 
-impl<'a, T: Source + ?Sized + 'a> Source for Box<T> {
-    fn source_id(&self) -> &SourceId {
-        (**self).source_id()
-    }
-
-    fn update(&mut self) -> CargoResult<()> {
-        (**self).update()
-    }
-
-    fn download(&mut self, id: &PackageId) -> CargoResult<Package> {
-        (**self).download(id)
-    }
-
-    fn fingerprint(&self, pkg: &Package) -> CargoResult<String> {
-        (**self).fingerprint(pkg)
-    }
-
-    fn verify(&self, pkg: &PackageId) -> CargoResult<()> {
-        (**self).verify(pkg)
-    }
+#[derive(Eq, Clone, Debug)]
+struct SourceIdInner {
+    url: Url,
+    canonical_url: Url,
+    kind: Kind,
+    // e.g. the exact git revision of the specified branch for a Git Source
+    precise: Option<String>,
+    registry_key: Option<String>,
```

Comment:
This might be best represented as a `Registry(String)` below in the `Kind` type, perhaps?

Reply:
This is how it was originally represented and I agree its cleaner but it required adding a custom impl of Hash and PartialEq for Kind, whereas this structure can piggyback the existing work to make two git refs normalize.

Possibly I'm just piling on the technical debt here (e.g. we should store these git specific fields in the Kind also), and I could refactor the whole thing a bit instead of doing it more.